src/twin/ssh_tunnel.rs

```diff
@@ -0,0 +1,449 @@
+            }
+            Ok(output) => {
+                if !output.status.success() {
+                    error!(
+                        "Failed to create ssh key pair: {}",
+                        str::from_utf8(&output.stderr).unwrap()
+                    );
+                    anyhow::bail!("Error on ssh key creation.");
+                }
+            }
+        }
+    }
+
+    match fs::read(pub_key_path.to_str().unwrap()) {
+        Ok(pub_key) => Ok(pub_key),
+        Err(err) => {
+            error!("Failed to open ssh public key: {}", err);
+            anyhow::bail!("Error reading ssh public key.");
+        }
+    }
+}
+
+// RAII handle to the temporary bastion host certificate
+struct CertificateFile {
+    path: PathBuf,
+}
+
+impl CertificateFile {
+    fn new(path: impl AsRef<Path>, data: &[u8]) -> Result<CertificateFile> {
+        let mut f = fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .mode(0o600)
+            .open(path.as_ref())?;
+        f.write_all(data)?;
+        Ok(CertificateFile {
+            path: path.as_ref().to_path_buf(),
+        })
+    }
+}
+
+impl Drop for CertificateFile {
+    fn drop(&mut self) {
+        if let Err(err) = fs::remove_file(&self.path) {
+            warn!(
+                "Failed to delete certificate \"{}\": {}",
+                self.path.to_str().unwrap(),
+                err
+            );
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SshTunnel {
+    state: FeatureState,
+}
+
+impl Feature for SshTunnel {
+    fn name(&self) -> String {
+        Self::ID.to_string()
+    }
+
+    fn version(&self) -> u8 {
+        Self::SSH_TUNNEL_VERSION
+    }
+
+    fn is_enabled(&self) -> bool {
+        env::var("SUPPRESS_SSH_TUNNEL") != Ok("true".to_string())
+    }
+
+    fn report_initial_state(&self) -> Result<()> {
+        self.report_ssh_tunnel_status()
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn state_mut(&mut self) -> &mut FeatureState {
+        &mut self.state
+    }
+
+    fn state(&self) -> &FeatureState {
+        &self.state
+    }
+}
+
+impl SshTunnel {
+    const SSH_TUNNEL_VERSION: u8 = 1;
+    const ID: &'static str = "ssh_tunnel";
+
+    pub fn refresh_ssh_tunnel_status(&self) -> Result<Option<serde_json::Value>> {
+        info!("ssh tunnel status requested");
+
+        self.ensure()?;
+
+        self.report_ssh_tunnel_status()?;
+
+        Ok(None)
+    }
+
+    pub fn get_ssh_pub_key(&self) -> Result<Option<serde_json::Value>> {
+        info!("ssh public key requested");
+
+        self.ensure()?;
+
+        self.report_ssh_tunnel_status()?;
+
+        #[derive(Serialize)]
+        struct PublicKeyResponse {
+            pub_key: Vec<u8>,
+        }
+
+        let response = PublicKeyResponse {
+            pub_key: create_or_get_pub_key()?,
+        };
+
+        Ok(Some(json!(response)))
+    }
+
+    pub fn report_ssh_tunnel_status(&self) -> Result<()> {
```

Comment:
did you run `cargo test -- --test-threads=1`?
We must discuss if connection state must additionally reported in twin since that's the way we do it with other features already...